src/transformers/trainer.py

```diff
@@ -1459,6 +1515,19 @@ def _save_checkpoint(self, model, trial, metrics=None):
         if self.is_world_process_zero():
             self._rotate_checkpoints(use_mtime=True, output_dir=run_dir)
 
+        # Save RNG state in non-distributed training
+        if self.is_local_process_zero():
+            rng_states = {"cpu": torch.random.get_rng_state()}
```

Comment:
I'm just wondering whether you checked that different processes in dist env will have the same RNG state?

Reply:
And another question - what about python's main RNG?

Reply:
In distributed training, the torch generator should have the same state on all processes since it's set with the same seed initially (then all processes execute the same code). Unless someone goes out of their way to only execute a random instruction on one process.

The code has been tested one GPU, 2 GPUs with DataParallel and 2 GPUs distributed and the same result are obtained with a full training and resuming from the last checkpoint.

As for python main RNG and numpy main RNG I have no way of extracting the seed (I can set it but not get the current state easily) so any code that wants to be 100% reproducible when resuming from a checkpoint needs to use torch RNGs only.

Reply:
> As for python main RNG and numpy main RNG I have no way of extracting the seed (I can set it but not get the current state easily)

Won't these work?
```
py_rng_state = random.getstate()
random.setstate(py_rng_state)

np_rng_state = numpy.random.get_state()
numpy.random.set_state(np_rng_state)
```

> so any code that wants to be 100% reproducible when resuming from a checkpoint needs to use torch RNGs only.

Do you mean to say that they should stick to `torch`'s random APIs? and convert to other formats from there?

Reply:
Oh TIL! I didn't think it was possible to extract and set those. Will add that to the PR.